### PR TITLE
rewrite of _utils._check_blocks _utils._check_same_gradients,

### PR DESCRIPTION
**Pull request by DavideTisi**
In the spirit of #119  I rewrote them in a more stringent way and allclose now use them to do all the checks for the metadata . This implies that all the test of allclose now are on these functions. I think everything has a test (maybe not all the gradient possibilities).

What do you think? 
I, of course, can create the equal functions, being the same as allclose except for the more stringent numerical checks, but I do not know if it really usefull.

### CODE REVIEW

**Comment by jwa7** [comment]
``_check_maps`` and ``_check_blocks`` seem like such fundamental functions that form the basis of most of our ``equistore.operations`` modules that I feel like they deserve their own dedicated and self-contained unit tests. Of course there would be some repetition in the unit tests of the public functions that use them, like ``add`` and ``allclose`` etc. 

Is it just a rule we follow that we don't test private functions at all? That all testing of them come from the public functions that wrap them?

File: python/src/equistore/operations/_utils.py [hunk 1]
```diff
@@ -48,58 +48,121 @@ def _check_blocks(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
...
+
+                err_msg_names = (
+                    f'gradient ("{parameter}") {prop} names are not the same '
+                    "or not in the same order."
+                )
+
+                if prop == "samples":
+                    if not len(grad_a.samples) == len(grad_b.samples):
+                        raise ValueError(err_msg + err_msg_len)
+                    if not grad_a.samples.names == grad_b.samples.names:
```
Comment (PicoCentauri) [inline python/src/equistore/operations/_utils.py:135]: Do I see this correct that we check for exactly the same gradients. If the values are shuffled in one gradients wrt to the second we throw an error. For me this is fine, but we should state this in the docstring.

File: python/src/equistore/operations/_utils.py [hunk 2]
```diff
@@ -48,58 +48,121 @@ def _check_blocks(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
...
+                        if not (c1.names == c2.names):
+                            raise ValueError(err_msg + err_msg_names)
+
+                        if not np.all(c1 == c2):
+                            raise ValueError(err_msg + err_msg_1)
+                elif prop == "parameters":
+                    # parameters are already checked at the begining but i want to give
+                    # the opportunity to the 'user' to use it, without problems
+                    continue
+                else:
```
Comment (PicoCentauri) [inline python/src/equistore/operations/_utils.py:160]: I think we can go for an `!=` here and leave the docstring.

```suggestion
                elif prop != "parameters":
                    # parameters are already checked at the begining but i want to give
                    # the opportunity to the 'user' to use it, without problems
```

**Review by PicoCentauri — CHANGES_REQUESTED** [review]
Good simplification here. I have some comments for the documentation and a possible simplification.


I would definetly vote for adding the public equal functions which will test these new functions properly.

File: python/src/equistore/operations/_utils.py (hunk 1 above)
Comment (DavideTisi) [reply python/src/equistore/operations/_utils.py:135]: Now these functions always check for the values and the order. All the operations that uses this check so far are implemented that need the order. Make them work non with the order is the next step

Comment (DavideTisi) [reply python/src/equistore/operations/_utils.py:135]: i added a comment in the docstring

**Review by PicoCentauri — APPROVED** [review]
Fine for me now. Can you rebase @DavideTisi

**Comment by DavideTisi** [comment]
did the rebase added the equal function

**Comment by DavideTisi** [comment]
with that we surpassed 100 tests 🎉🎉

**Review by jwa7 — APPROVED** [review]
Looks good!

File: python/src/equistore/operations/_utils.py [hunk 3]
```diff
@@ -48,58 +48,119 @@ def _check_blocks(a: TensorBlock, b: TensorBlock, props: List[str], fname: str):
...
+
+    If props is None it only checks if the ``'parameters'`` are consistent.
+
+    :param a: first :py:class:`TensorBlock` for check
+    :param b: second :py:class:`TensorBlock` for check
+    :param props: A list of strings containing the property to check.
+                 Allowed values are ``'samples'`` or ``'properties'``,
+                 ``'components'``. To check only if the ``'parameters'`` are consistent
+                 use ``props=None``, using ``'parameters'`` is allowed
+                  even though deprecated.
```
Comment (Luthaf) [inline python/src/equistore/operations/_utils.py:102]: Why are we keeping around deprecated stuff? We don't have a release yet, we can break everything =)

File: python/src/equistore/operations/equal.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def equal(tensor1: TensorMap, tensor2: TensorMap, only_metadata: bool = False) -> bool:
+    """Compare two :py:class:`TensorMap`.
+
+    This function returns ``True`` if the two tensors have the same keys
+    (potentially in different order) and all the :py:class:`TensorBlock` have
+    the same (and in the same order) samples, components, properties,
+    and their values matrices pass the == test with the provided
+    ``rtol``, and ``atol``.
```
Comment (Luthaf) [inline python/src/equistore/operations/equal.py:16]: There is no `atol`/`rtol` in this function parameters

File: python/src/equistore/operations/_utils.py (hunk 3 above)
Comment (DavideTisi) [reply python/src/equistore/operations/_utils.py:102]: deprecated is maybe not the correct word, the idea is that I wanted to have both behavior: `props=None` -> checks the ``parameters`` and  the possibility to not break the code for someone that writes: `props=['samples',... whatever...,'parameters']`, thinking that `parameters` is needed to check the parameters, which are actually always checked.

File: python/src/equistore/operations/equal.py (hunk 4 above)
Comment (DavideTisi) [reply python/src/equistore/operations/equal.py:16]: it is a copy-paste error

Comment (DavideTisi) [reply python/src/equistore/operations/equal.py:16]: i will correct it thanks

File: python/src/equistore/operations/_utils.py (hunk 3 above)
Comment (Luthaf) [reply python/src/equistore/operations/_utils.py:102]: I would raise an error if someone passes `parameters`, and make it clear in the doc that they are always checked

**Pull request merged**